Mathlib/Tactic/AddTacticDoc.lean

```diff
@@ -0,0 +1,149 @@
+    ("description", description)
+  ]
+
+/--
+Serves as a global storage for all `TacticDocEntry`s we have added throughout mathlib codebase.
+When [doc-gen](https://github.com/leanprover-community/doc-gen) generates documentation, it will
+access this storage to find each `tacticDocEntry` along with its greatest-priority docstring.
+-/
+initialize docNamesExtension : SimplePersistentEnvExtension
+  (TacticDocEntry × String) (List (TacticDocEntry × String)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := λ _ => {}
+    addEntryFn    := λ xs x => x :: xs
+  }
+
+/--
+A command used to add documentation for a tactic, command, hole command, or attribute.
+
+Usage: after defining an interactive tactic, command, or attribute,
+add its documentation as follows.
+```lean
+/--
+describe what the command does here
+-/
+add_tactic_doc
+{ name := "display name of the tactic",
+  category := cat,
+  decl_names := [`dcl_1, `dcl_2],
+  tags := ["tag_1", "tag_2"] }
+```
+
+The argument to `add_tactic_doc` is a structure of type `TacticDocEntry`.
+* `name` refers to the display name of the tactic; it is used as the header of the doc entry.
+* `category` refers to the category of doc entry.
+  Options: `doc_category.tactic`, `doc_category.cmd`, `doc_category.hole_cmd`, `doc_category.attr`
+* `decl_names` is a list of the declarations associated with this doc. For instance,
+  the entry for `linarith` would set ``decl_names := [`tactic.interactive.linarith]``.
+  Some entries may cover multiple declarations.
+  It is only necessary to list the interactive versions of tactics.
+* `tags` (optional) is a list of strings used to categorize entries.
+* `inherit_description_from` (optional) is a name of the declaration that contains a docstring.
+
+Now, the body of the doc entry is determined as follows:
+* 1st priority: the docstring on top of the `add_tactic_doc` command call itself.
+* 2nd priority: the docstring of the declaration from the `inherit_description_from` field.
+* 3rd priotiy: if none of the above were found, and if `decl_names` contains
+  one and only one declaration, we get the docstring from that declaration.
+
+So, if there are multiple declarations, you can select the one to be used
+by passing a name to the `inherit_description_from` field.
+If you prefer your tactic/command/etc. to have a docstring that is different from
+the online documentation entry, you should write the doc entry as a docstring for
+the `add_tactic_doc` invocation.
+
+Note that providing a badly formed `tactic_doc_entry` to the command can result in strange error
+messages.
+-/
+elab metaDocString:docComment ? "add_tactic_doc " tdeStx:term : command => do
+  -- 1. Turn tde:TSyntax argument into the actual tde:TacticDocEntry object
+  let tdeName : Name ← resolveGlobalConstNoOverloadWithInfo tdeStx
+  let tde : TacticDocEntry ← unsafe evalConstCheck TacticDocEntry ``TacticDocEntry tdeName
```

Comment:
This `unsafe` makes me nervous. Is there a way to avoid it?

Reply:
This particular pattern is fairly normal, and I would expect it to come up in something like this where we want to make an actual `TacticDocEntry` from a syntax representing such. This is automated to some extent by `declare_config_elab` but that is specialized for config arguments in tactics.